### PR TITLE
[installer-tests] Use reference architecture to run integration tests

### LINKED ISSUE #12865: [terraform] Use GCP and AWS reference architecture to run automated tests

### PR DESCRIPTION
**Pull request by nandajavarma**
## Description

This PR changes the `installer-tests.ts` and the `Makefile` being called from it, to use the single-cluster reference architecture to run the integration tests. This is the third of a 3 part PR series. This PR requires the other two PRs to be gone in first and that would reduce the file changes in this file to very few in number.

The two preceding PRs are: **TO BE REVIEWED BEFORE REVIEWING THIS PR**
- [ ] https://github.com/gitpod-io/gitpod/pull/13540
- [ ] https://github.com/gitpod-io/gitpod/pull/13561

The following are the test runs:
k3s: https://werft.gitpod-dev.com/job/gitpod-custom-nvn-add-tf-tests.31 :heavy_check_mark:
gke: https://werft.gitpod-dev.com/job/gitpod-custom-nvn-add-tf-tests.32 :heavy_check_mark:
aks: https://werft.gitpod-dev.com/job/gitpod-custom-nvn-add-tf-tests.33 :heavy_check_mark:
eks: https://werft.gitpod-dev.com/job/gitpod-custom-nvn-add-tf-tests.34 :heavy_check_mark:

## Related Issue(s)

Fixes #12865

## How to test

You can follow this [internal](https://www.notion.so/gitpod/WIP-Self-hosted-automated-nightly-pipelines-fdc5a202e67b4197aa00d93b98d57927) doc to trigger self-hosted setups.

## Release Notes

```release-note
NONE
```

## Werft options:

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [ ] /werft with-preview
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`

### CODE REVIEW

File: install/infra/modules/aks/local.tf
Comment (adrienthebo) [inline install/infra/modules/aks/local.tf:17]: **thought(style):** while slipping in style fixes in smaller pull requests is no big deal, this is a pretty hefty pull request and style fixups can increase the cost of reviewing. I'll absolutely rubber stamp pull request that just runs code reformatting so if you're in a situation where code needs to be reformatted consider extracting a quick format PR so we can iterate faster on the fun stuff. 😊

File: install/infra/modules/k3s/dns.tf
Comment (adrienthebo) [inline install/infra/modules/k3s/dns.tf:15]: **question:** Do we anticipate the need to use a pre-existing Cloud DNS zone? The release preview environments (eg `release-<xxx>.tests.gitpod-self-hosted.com`) domains use a pre-existing domain, will this code work when we stand up preview environments where the Cloud DNS zone is managed outside of this module?

File: install/infra/modules/k3s/local.tf
Comment (adrienthebo) [inline install/infra/modules/k3s/local.tf:1]: **opinion(non-blocking):** This is purely my opinion and deviates from the standards, but I personally like having local variables declared closer to the place where they're being used - if I can see all the context in a single pane in my editor then I need to do less chasing across a module. Purely subjective opinion on my part, no action needed.

File: install/infra/modules/tools/cloud-dns-ns/main.tf
Comment (adrienthebo) [inline install/infra/modules/tools/cloud-dns-ns/main.tf:9]: **issue(non-blocking):** If we're anticipating that external users will be consuming our modules then we can't hardcode the names of our projects in `install/infra/modules/*`.

Would the following work? (Not sure if it's even valid syntax, just shooting from the hip)

```suggestion
  project     = var.dns_project
```

Non-blocker, feel free to skip this issue and address later.

File: install/tests/Makefile
Comment (adrienthebo) [inline install/tests/Makefile:12]: 🤩

File: install/tests/Makefile
Comment (adrienthebo) [inline install/tests/Makefile:7]: **question:** We're silencing Make's default behavior to print commands before running them. Could this make tests more difficult to debug? Does this improve the overall user experience? Who benefits the most/the least from silencing that output?

File: install/tests/Makefile
Comment (adrienthebo) [inline install/tests/Makefile:172]: **suggestion(non-blocking, out-of-scope):** We reactivate service account credentials in a number of makefile targets, and also diffuses the responsibility of providing credentials across a bunch of different make targets. What do you think of having a make target like `activate-credentials` that runs at the beginning of all our test configurations? Moving this logic to a single target and using `installer-tests.ts` to execute the makefile target ensures we have a consistent environment for the rest of the test execution, decouples environment setup from the individual tests, and gives us a clear point where a user manually running make targets can opt-in (or opt-out) of this activation behavior.

Non-blocker, no change required; out of scope; this is a non-trivial change that we can address in another pull request.

File: install/tests/Makefile
Comment (adrienthebo) [inline install/tests/Makefile:182]: **question:** Under which conditions would `make get-aws-secret` fail to generate a credentials file, but touching the credential file would let the rest of test execution work? I ran into this when chasing down some other failures and didn't understand what's going on; if there's some subtle behavior at play can we document what's going in via a comment?

File: install/tests/Makefile
Comment (adrienthebo) [inline install/tests/Makefile:387]: **issue(out-of-scope):** We're implementing optional behavior by doing `optional-command || echo "..."`; this works but when debugging failed CI jobs I've struggled to understand if these messages are indicating a problem. I would find it really helpful to amend these messages to indicate what's going on, if it's a problem, and if I need to do anything when encountering it.

File: install/tests/Makefile
Comment (adrienthebo) [inline install/tests/Makefile:385]: **issue(bug):** I'm pretty sure that this will never backup the gitpod certificates; we delete the gitpod app _before_ attempting to backup the cert. Is this intended?

**Review by adrienthebo — COMMENTED** [review]
No blockers, some questions, and some small issues. After getting a bit more familiar with the automated testing code I've managed to exercise a bunch of the weird cases by misusing any number of things 😅 so I'm adding a few more comments there; we'll have other people within Gitpod as well as outside using these more and more and we'll be gradually shifting to more polish type questions as more people touch this code. PTAL; let me know when you've done another pass and I'll take a look but expect to approve this on the next pass. ⚡️

File: install/tests/Makefile
Comment (nandajavarma) [reply install/tests/Makefile:385]: Deleting the gitpod app doesn't cleanup the namespace. the secret will still be there.

Comment (nandajavarma) [reply install/tests/Makefile:385]: please see this PR  that fixes this issue out of the scope of this https://github.com/gitpod-io/gitpod/pull/13533

File: install/tests/Makefile
Comment (nandajavarma) [reply install/tests/Makefile:387]: We can take this up in another improvement PR

File: install/tests/Makefile
Comment (nandajavarma) [reply install/tests/Makefile:182]: It would fail if there is no such credentials file present in GCS. We are not catching this error anymore. This would fail [here](https://github.com/gitpod-io/gitpod/pull/13542/files/021b9bf50683f76951eedf3f191f683a23b3a73e#diff-bd7b7f3a397df93111a6b6e2a4284b01bd57e04474bc9f91f1b02ee1960399a2R174) with the error

File: install/tests/Makefile
Comment (nandajavarma) [reply install/tests/Makefile:7]: I think silencing the commands itself makes it much easier to debug since we can actually see on the error messages. With this change the logs become much more readable. Alongside the user who will be getting kubeconfig will have a smoother experience.

File: install/infra/modules/tools/cloud-dns-ns/main.tf
Comment (nandajavarma) [reply install/infra/modules/tools/cloud-dns-ns/main.tf:9]: Will take this change up on the [corresponding PR](https://github.com/gitpod-io/gitpod/pull/13541)

File: install/infra/modules/k3s/dns.tf
Comment (nandajavarma) [reply install/infra/modules/k3s/dns.tf:15]: This is not a use case we anticipate. Currently in GCP, AWS and Azure the behavior is that we create a managed zone by the name and add the NS record. I am only making the behavior consistent here.

File: install/infra/modules/aks/local.tf
Comment (nandajavarma) [reply install/infra/modules/aks/local.tf:17]: Sorry for the confusion here. I have kind of accumulated all the lint changes on [this PR ](https://github.com/gitpod-io/gitpod/pull/13541), I should have put this PR on draft to avoid the risk of it getting reviewed.

**Comment by nandajavarma** [comment]
Hey @adrienthebo ! Thanks a lot for the thorough review on this. I am sorry if I was not clear enough, but this PR depends on https://github.com/gitpod-io/gitpod/pull/13541 and https://github.com/gitpod-io/gitpod/pull/13541 to go in. Which would reduce the size of this one massively. The problem you mentioned about having linting changes would be resolved since the second PR is basically just the linting changes. The first PR is all the changes made to k3s terraform module. I can move this PR to draft if you would like. I will make the changes you suggested on the [k3s PR ](https://github.com/gitpod-io/gitpod/pull/13541).

**Comment by nandajavarma** [comment]
Moving to Draft until https://github.com/gitpod-io/gitpod/pull/13561 and https://github.com/gitpod-io/gitpod/pull/13533 are merged

**Review by adrienthebo — APPROVED** [review]
LGTM, but there's an outstanding merge conflict. I'm preemptively approving this to unblock you; let's attach a green CI run to the PR description prior to merging. 😊

/hold

**Comment by nandajavarma** [comment]
Running jobs:
k3s: https://werft.gitpod-dev.com/job/gitpod-custom-nvn-add-tf-tests.31
gke: https://werft.gitpod-dev.com/job/gitpod-custom-nvn-add-tf-tests.32
aks: https://werft.gitpod-dev.com/job/gitpod-custom-nvn-add-tf-tests.33
eks: https://werft.gitpod-dev.com/job/gitpod-custom-nvn-add-tf-tests.34

**Comment by nandajavarma** [comment]
Will hold the PR until the release goes

**Pull request merged**
